File: src/ts/sort/mergesort.ts
// QuickSort sorts array of items
export function mergeSort(arr: number[]) {
	msort(arr, 0, arr.length - 1)
}

function msort(arr: number[], low: number, high: number) {
	if (low < high) {
    const mid = Math.floor((low + high) / 2)
		msort(arr, low, mid)
    msort(arr, mid+1, high)
    merge(arr, low, mid, high)
	}
}

function merge(arr: number[], low: number, mid: number, high: number) {
  const tmplen = mid - low + 1
  const tmp = arr.slice(low, mid + 1)

	let d = low
	let i = 0
	let j = mid + 1
	for (; i < tmplen && j <= high; d++) {
		if (tmp[i] < arr[j]) {
			arr[d] = tmp[i]
			i++
		} else {
			arr[d] = arr[j]
			j++
		}
	}

	while (i < tmplen) {
		arr[d] = tmp[i]
		i++
		d++
	}
}

File: go/src/sort/mergesort.go
package sort

// MergeSort sorts array of items arr
func MergeSort(arr []Item) {
	msort(arr, 0, len(arr)-1)
}

func msort(arr []Item, low, high int) {
	if low < high {
		mid := (low + high) / 2
		msort(arr, low, mid)
		msort(arr, mid+1, high)
		merge(arr, low, mid, high)
	}
}

func merge(arr []Item, low, mid, high int) {
	tmplen := mid - low + 1
	tmp := make([]Item, tmplen)
	copy(tmp, arr[low:mid+1])

	d := low
	i := 0
	j := mid + 1
	for ; i < tmplen && j <= high; d++ {
		if tmp[i].Less(arr[j]) {
			arr[d] = tmp[i]
			i++
		} else {
			arr[d] = arr[j]
			j++
		}
	}

	for i < tmplen {
		arr[d] = tmp[i]
		i++
		d++
	}
}
